### ISSUE TITLE
[aws] Avoid writing secrets to user-data

### ISSUE THREAD

**Issue by captn3m0** [Quentin-M/etcd-cloud-operator#31]
https://github.com/Quentin-M/etcd-cloud-operator/blob/78eee7e1bc1837af58b95c56da3a1858859c248b/terraform/modules/ignition/ignition.tf#L121-L129

I was trying to think of ways to get the key pushed to the instance without writing it via user-data (AWS-only). The concern stems from people/tools having access to the [DescribeInstanceAttribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceAttribute.html) API call which will gladly return this back.

The possible approach I was considering was:

1. Write the secret to a [secret version](https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret_version.html) resource.
2. Add a new flag on the eco binary to instead fetch the secret from the secretmanager instead of reading from disk.

For (2), another option is to add a secret-fetch systemd service which does this, but then figuring out how to get aws cli running on the instance is another problem.

Thoughts?

**Comment by captn3m0** [comment]
Another alternative is to write it to S3, and fetch it from S3 via the `ignition_file` source argument, which supports S3: https://www.terraform.io/docs/providers/ignition/d/file.html#source-1

**Comment by Quentin-M** [comment]
Hi @captn3m0,

You are absolutely right. It's yet another of my oversight from the time I was POC'ing this project & never thought of coming back to this, unfortunately.

S3 support for Ignition is something I pushed CoreOS hard to implement back in Spring 2017 when I was still working there actually. 

It should be quite straightforward to rely on the S3 ACLs / EC2 IAM instance profile as well as the server-side encryption + TLS upload for S3 (except if you want to have your own keys - which is fine too).

**Comment by captn3m0** [comment]
The extra resources created are:

1. S3 bucket for storing etcd secrets
2. IAM resources to the secrets
3. S3 object storing the secrets
4. KMS Key for encrypting secrets to S3. I want to ensure that this uses a new key created by this module.

If we're okay with pushing stuff to S3, an easier solution would be to push the entire ignition config to S3 instead:

```
resource "aws_s3_bucket_object" "ignition_master" {
  bucket                 = "bucket_name"
  key                    = "path/${random_id.ignition-key.hex}.json"
  content                = "${data.ignition_config.main.rendered}"
  acl                    = "private"
  server_side_encryption = "AES256"
}

resource "random_id" "ignition-key" {
  byte_length = 8
}

data "ignition_config" "s3" {
  replace {
    source       = "${format("s3://%s/%s", "bucket_name", aws_s3_bucket_object.ignition_master.key)}"
    verification = "sha512-${sha512(data.ignition_config.main.rendered)}"
  }
}


// ignition_config.main goes here
```

Okay with either approach, whichever you think works better. The question I had was with concerns around the creation of the bucket: Should this be in the module here?

I haven't dabbled with the kubernetes module yet, would that be affected by this change?

**Comment by Quentin-M** [comment]
I agree with you, this is totally appropriate. We could definitely push the entire configuration to S3 - this is actually extremely straightforward as well. Using our own KMS key as you described, even better.

There is also the fact that the S3 provider pushes the etcd data backups to S3, and they should also be encrypted fully. Also, the TF state stored in S3.

Feel free to create the new bucket in [here](https://github.com/Quentin-M/etcd-cloud-operator/blob/b0137cf15c9c2ce30e580a8db4c1c79c7a619af3/terraform/platforms/aws/aws.tf#L15), alongside the `backup` one. With the KMS configuration, as well as the IAM role - it may make sense to create a new `s3.tf` file to hold everything together.

Regarding using a `random_id` to compute the filename, I imagine that you would be doing this to somehow version the configuration? If so, this is not necessary if you enable versioning within the S3 bucket itself. New nodes coming up will always be able to use the latest configurable available in S3, without requiring to modify the ignition_config necessarily. (Also because this `random_id` resource is only randomized once.) Except if you want to leverage ASG configuration changes to automatically perform rolling updates of the etcd nodes, but I would be a little bit careful with that - as not only does a node need to be up, but etcd needs to have recovered fully (which can take some time).

In your S3 object resource above, you need to be using those parameters:
```
- server_side_encryption - (Optional) Specifies server-side encryption of the object in S3. Valid values are "AES256" and "aws:kms".
- kms_key_id - (Optional) Specifies the AWS KMS Key ARN to use for object encryption. This value is a fully qualified ARN of the KMS Key. If using aws_kms_key, use the exported arn attribute: kms_key_id = "${aws_kms_key.foo.arn}"
```

Cheers!

**Comment by captn3m0** [comment]
Got it, the `random_id` based versioning was from another codebase where we use this at multiple clusters. Won't be needed here :+1: 

Will file a PR soon.

**Comment by captn3m0** [comment]
TIL: ignition logs the entire configuration to journalctl

```Feb 13 05:08:37 localhost ignition[417]: parsing config: {"ignition":{"config":{"replace":{"source":"s3://eco-example-config/config.json","verification":{"has>
Feb 13 05:09:49 localhost ignition[417]: fetched referenced config: {"ignition":{"config":{"append":[{"source":"data:text/plain;charset=utf-8;base64,eyJpZ25pd>
Feb 13 05:09:49 localhost ignition[417]: fetched referenced config: {"ignition":{"config":{},"timeouts":{},"version":"2.1.0"},"networkd":{},"passwd":{},"stora>
Feb 13 05:09:49 localhost ignition[417]: disks: op(1): [started]  waiting for udev to settle
```

**Comment by Quentin-M** [comment]
Oh boy, great find. That is fantastic.. I was not even aware of that when I
worked at CoreOS.. I would assume (aka hope so) that when the entire
ignition config is stored in S3 (besides the append), the logs would only
show the S3 URL?

-- 
Quentin Machu

**Comment by captn3m0** [comment]
No, it logs every `ignition_config` block that it fetches, even those from external sources: https://github.com/coreos/ignition/blob/aad24ad59393d49d1e7cdf6c4504a94615d9f0c3/internal/exec/engine.go#L264

Any idea how I can reduce the verbosity for this to avoid this getting logged?

**Comment by captn3m0** [comment]
Filed an issue on Ignition: https://github.com/coreos/ignition/issues/721

**Comment by Quentin-M** [comment]
In the [docs](https://coreos.com/ignition/docs/latest/getting-started.html)
they mention increasing the verbosity so we should be able to reduce it the
same way. Seems a bit abusive to change the global parameter like that
though.

Technically, ignition only logs to stdout, so whatever application collect
ignition's stdout should be re-configured. What we need to check is that we
actually are not actually collecting them - and thus storing them somewhere
then filtering based on config later.

-- 
Quentin Machu

**Comment by Quentin-M** [comment]
I am wondering if you should take this public issue down and report it privately first instead. This could technically be an exploitable vulnerability here, as logs could be read by local non-root users and used to take over private keys and whatnot.

**Comment by Quentin-M** [comment]
The joy of Github keeping every single bit of the history.. You can still remove the old `Edited` content in the submenu.

![image](https://user-images.githubusercontent.com/1332289/52701461-21e2f080-2f2f-11e9-8f59-3fe7fc6ff210.png)

**Comment by lucab** [comment]
If I remember correctly, non-privileged users should not be able to read system log via `journalctl`. On ContainerLinux, both the `root` and `core` users are allowed to do that via group permissions.

**Comment by Quentin-M** [comment]
@lucab Looks accurate indeed - saves us from random container escapes. The only last bit would be that lots of enterprise customers ship their logs to a central server - so it'd depend on authz there. Ops should then make sure that ContainerLinux's syslogs are considered sensitive.

**Comment by captn3m0** [comment]
So, this was fixed on ignition 0.31.0, which needs a minimum of the following CoreOS releases:

- stable: 2079.3.0 
- beta: 2079.1.0 
- alpha: 2065.0.0
